src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,411 @@
+                # TODO(meredith): log info for the different errors
+                # that could happen:
+                # * CancelledError (future was cancelled)
+                # * TimeoutError (future timedout)
+                # * Exception (the fucture call raised an exception)
+                raise
+            self.__commit_function({message.partition: Position(message.offset, message.timestamp)})
+
+    def submit(self, outer_message: Message[MessageBatch]) -> None:
+        assert not self.__closed
+
+        if len(self.__futures) >= self.__max_buffer_size:
+            raise MessageRejected
+
+        for message in outer_message.payload:
+            payload = message.payload
+            future = self.__producer.produce(
+                destination=Topic(self.__producer_topic),
+                payload=payload,
+            )
+            self.__futures.append(ProducerResultFuture(message, future))
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+        self.__producer.close()
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+        while self.__futures:
+            remaining = timeout - (time.time() - start) if timeout is not None else None
+            if remaining is not None and remaining <= 0:
+                logger.warning(f"Timed out with {len(self.__futures)} futures in queue")
+                break
+
+            message, future = self.__futures.popleft()
+
+            future.result(remaining)
+
+            self.__commit_function({message.partition: Position(message.offset, message.timestamp)})
+        self.__producer.close()
+
+
+def process_messages(
+    outer_message: Message[MessageBatch],
+) -> MessageBatch:
+    """
+    We have an outer_message Message() whose payload is a batch of Message() objects.
+
+        Message(
+            partition=...,
+            offset=...
+            timestamp=...
+            payload=[Message(...), Message(...), etc]
+        )
+
+    The inner messages payloads are KafkaPayload's that have:
+        * key
+        * headers
+        * value
+
+    The value of the message is what we need to parse and then translate
+    using the indexer.
+    """
+    indexer = get_indexer()
+    metrics = get_metrics()
+    task = get_task()
+
+    strings = set()
+    parsed_payloads_by_offset = {
+        msg.offset: json.loads(msg.payload.value.decode("utf-8"), use_rapid_json=True)
+        for msg in outer_message.payload
+    }
+
+    for message in parsed_payloads_by_offset.values():
+        metric_name = message["name"]
+        tags = message.get("tags", {})
+
+        parsed_strings = {
+            metric_name,
+            *tags.keys(),
+            *tags.values(),
+        }
+        strings.update(parsed_strings)
+
+    with metrics.timer("metrics_consumer.bulk_record"):
+        mapping = indexer.bulk_record(list(strings))  # type: ignore
```

Comment:
Why does this have to ignore types ?
Also why passing a list to `bulk_record` It seems it should take a set

Reply:
@fpacifici this was actually already like this, I remember initially passing in a set, and then changed to list I don't remember why. I could change it back to a set tho

Reply:
not in this PR